constph.py

```diff
@@ -732,7 +733,98 @@ def setNumAttemptsPerUpdate(self, nattempts=None):
             # TODO: Cache already-visited states to avoid recomputing?
             self.nattempts_per_update = self.getNumTitratableGroups()
-        
+
+
+class CalibrationTitration(MonteCarloTitration):
+    """Implementation of self-adjusted mixture sampling for calibrating titratable residues.
+
+
+
+    """
+    def __init__(self, system, temperature, pH, prmtop, cpin_filename, nattempts_per_update=None, simultaneous_proposal_probability=0.1, target_weights=None, debug=False):
+        """
+        target_weights (list) - None or nested list indexed[group][state] of relative weights (pi) for sams method
+        """
+
+        super(CalibrationTitration, self).__init__(system, temperature, pH, prmtop, cpin_filename, nattempts_per_update=None, simultaneous_proposal_probability=0.1, debug=False)
+
+        self.n_updates=0
+
+        for i,group in enumerate(self.titrationGroups):
+            for j, state in enumerate(self.titrationGroups[i]['titration_states']):
+                if target_weights is not None:
+                    self.titrationGroups[i]['titration_states'][j]['target_weight'] = target_weights[i][j]
+                else:
+                    self.titrationGroups[i]['titration_states'][j]['target_weight'] = 1.0 / len(self.titrationGroups[i]['titration_states'])
+
+    def update(self, context,scheme=None):
+        super(CalibrationTitration, self).update(context)
+        self.n_updates += 1
+
+        if scheme is not None:
+            temperature = self.temperature
+            kT = kB * temperature  # thermal energy
+            beta = 1.0 / kT  # inverse temperature
+            zeta = numpy.asarray(map(lambda x: x['relative_energy'], self.titrationGroups[0]['titration_states'][:]))
+            zeta *= -beta  # zeta^{t-1}
+            if scheme in ['theorem1', 'eq9']:
+                update = self._theorem1()
+            elif scheme in ['theorem2', 'eq12']:
+                update = self._theorem2(context, beta, zeta)
+
+            # zeta^{t-1/2}
+            zeta += update
+            # zeta^{t} = zeta^{t-1/2} - zeta_1^{t-1/2}
+            zeta_t = zeta - (numpy.ones_like(zeta) * zeta[0])
+
+            # Set reference energy based on new zeta
+            for i, titr_state in enumerate(zeta_t):
+                self.titrationGroups[0]['titration_states'][i]['relative_energy'] = titr_state / -beta
+
+
+    def _theorem1(self):
+        # [1/pi_1...1/pi_i]
+        update = numpy.asarray(map(lambda x: 1 / x['target_weight'], self.titrationGroups[0]['titration_states'][:]))
+        # delta(Lt)
+        delta = numpy.zeros_like(update)
+        delta[self.getTitrationState(0)] = 1
+        update *= delta
+        update /= self.n_updates  # t^{-1}
+        return update
+
+    def _theorem2(self, context, beta, zeta):
+        # target weights
+        pi_j = numpy.asarray(map(lambda x: x['target_weight'], self.titrationGroups[0]['titration_states'][:]))
```

Comment:
Yes, right now they're set using 
```python
else:
                    self.titrationGroups[i]['titration_states'][j]['target_weight'] = 1.0 / len(self.titrationGroups[i]['titration_states'])
```
And they're both 0.5. I use the same weights for the original theorem and it works.